Homework5

Kiba23 opened the pull request.

Kiba23 committed:
Init commit.

Just initialized the projects, branch and solutions.

Kiba23 committed:
Completed Task1 in Homework5.

Implemented all the things, all works good. Provided tests of all the features in Program class.

Kiba23 committed:
Completed the Task2.

Implemented all the things. All works good. Provided tests in Program class.

Kiba23 committed:
Completed the Task2 in Homework5.

Changed some fields accessors from previous commit. All logic remains the same, all works.

Kiba23 committed:
Init the Task3 in Homework5.

Just initialized classes in Task3 Homework5.

Kiba23 committed:
Completed Task3 and whole Homework5.

From init till finish Task3. All works. Provided Tests in Program class. Comments included.

radislavkireev commented on Homework5/Employees/Models/Organization.cs:
```diff
@@ -0,0 +1,81 @@
+            {
+                result += vacation.VacationPeriod;
+            }
+            return result / Vacations.Count;
```
check LINQ and user Average method: https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.average?view=net-6.0

radislavkireev commented on Homework5/Employees/Models/Organization.cs:
```diff
@@ -0,0 +1,81 @@
+                    summedVacationTime += vacation.VacationPeriod;
+                }
+                yield return (person.Key, summedVacationTime / person.Count());
+                summedVacationTime = default(TimeSpan);         // setting sum to zero again, for a new person
```
Please use LINQ methods for calculations and selection

radislavkireev commented on Homework5/ISBNbook/Models/Book.cs:
```diff
@@ -0,0 +1,33 @@
+            }
+            PublicationDate = publicationDate;
+            Authors = new List<string>();
+            Authors.AddRange(authors);
```
Authors should not be empty and should be unique

radislavkireev commented on Homework5/ISBNbook/Models/Catalogue.cs:
```diff
@@ -0,0 +1,69 @@
+        {
+            if (IsFormatCorrect(isbn))
+            {
+                BookCollection.Add(ToDefaultFormat(isbn), book);
```
The code will throw exception when the book with the same isbn is added, consider to update the existing book intstead

radislavkireev commented on Homework5/ISBNbook/Models/Catalogue.cs:
```diff
@@ -0,0 +1,69 @@
+{
+    public class Catalogue
+    {
+        private Dictionary<string, Book> BookCollection = new Dictionary<string, Book>();
```
lowercase private field name

radislavkireev commented on Homework5/SparseMatrix/Entities/SparseMatrix.cs:
```diff
@@ -0,0 +1,127 @@
+{
+    public class SparseMatrix : IEnumerable<int>
+    {
+        private Dictionary<(int Column, int Row), int> MatrixElements = new Dictionary<(int, int), int>();
```
make private field names lowercase

radislavkireev commented on Homework5/SparseMatrix/Entities/SparseMatrix.cs:
```diff
@@ -0,0 +1,127 @@
+                InputDimensionCheck(col, row);
+
+                var elementIndex = (col, row);
+                MatrixElements.Add(elementIndex, value);
```
This setter should not store 0 values.
Assume we have a sparse matrix 1 mln x 1mln elements and it has single non-zero element.
Current code will take memory of size 10^12 bytes * N, where N is size of 1 element.

radislavkireev replied:
Also, consider the case, when element with the same index updates a few times:
```
var sparseMatrix  = new SparseMatrix();
sparseMatrix[0,0] = 0;
sparseMatrix[0,0] = 1;
```

radislavkireev commented on Homework5/SparseMatrix/Entities/SparseMatrix.cs:
```diff
@@ -0,0 +1,127 @@
+                }
+
+                var elementIndex = (col, row);
+                return MatrixElements.First(e => e.Key.Equals(elementIndex)).Value; // returning the value at corresponding indexes
```
access an element via Dictonary getter, do not call .First, it iterates over the all element under the hood.

radislavkireev commented on Homework5/SparseMatrix/Entities/SparseMatrix.cs:
```diff
@@ -0,0 +1,127 @@
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            for (int i = 0; i < Size.Columns; i++)
```
repetitive code in enumerators, refactor to remove duplication

Kiba23 committed:
Fixed HW5 after Radislav's suggestions.

Also added HW4 folder to homework5 branch, because I accidently removed it.
All works good. Provided test in Program class.
